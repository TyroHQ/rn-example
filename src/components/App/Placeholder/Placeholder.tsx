import React from "react";
import { Text } from "react-native";
import styled from "styled-components/native";

const PlaceholderContainer = styled.View<{ orientation: "row" | "column" }>`
  border: 1px solid grey;
  padding: 10px;
  display: flex;
  flex-direction: ${props => props.orientation};
  align-items: center;
  justify-content: center;
  flex: 1;
`;

export class Placeholder extends React.Component<{
  children?: React.ReactNode;
  orientation?: "row" | "column";
}> {
  static defaultProps: Placeholder["props"] = {
    orientation: "row"
  };

  render() {
    return (
      <PlaceholderContainer orientation={this.props.orientation}>
        {typeof this.props.children === "string" ? (
          <Text>{this.props.children}</Text>
        ) : (
          this.props.children
        )}
      </PlaceholderContainer>
    );
  }
}
