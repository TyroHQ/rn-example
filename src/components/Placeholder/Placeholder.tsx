import React from "react";
import { Text, View } from "react-native";
import styled from "styled-components";

const PlaceholderContainer = styled(View)<{ orientation: "row" | "column" }>`
  border: 1px solid grey;
  padding: 10px;
  display: flex;
  flex-direction: ${props => props.orientation};
  align-items: center;
  justify-content: center;
  flex: 1;
`;

function wrapString(c: React.ReactNode, i?: number): React.ReactNode {
  if (typeof c === "string") return <Text key={i}>{c}</Text>;

  if (React.isValidElement(c)) return React.cloneElement(c, { key: i });
  return c;
}

export class Placeholder extends React.Component<{
  children?: React.ReactNode;
  orientation?: "row" | "column";
}> {
  static defaultProps: Placeholder["props"] = {
    orientation: "row"
  };

  render() {
    const children = Array.isArray(this.props.children)
      ? this.props.children.map(wrapString)
      : wrapString(this.props.children);
    return (
      <PlaceholderContainer orientation={this.props.orientation}>
        {children}
      </PlaceholderContainer>
    );
  }
}
