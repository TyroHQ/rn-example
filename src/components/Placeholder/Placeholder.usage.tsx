import React from "react";
import { boolean, text, number, selectV2 } from "@storybook/addon-knobs";
import { Placeholder } from "./";

export const basic = () => {
  // const t = text("Content", "This can be replaced");
  const t = "This can be replaced";
  return (
    <Placeholder
      orientation="row"
      // orientation={selectV2(
      //   "Orientation",
      //   { row: "row", column: "column" },
      //   "row"
      // )}
    >
      <Placeholder>{t}</Placeholder>
      <Placeholder>{t}</Placeholder>
      <Placeholder>{t}</Placeholder>
    </Placeholder>
  );
};

export const withNumber = () => <Placeholder>{5}</Placeholder>;
export const withString = () => <Placeholder>Hello</Placeholder>;
