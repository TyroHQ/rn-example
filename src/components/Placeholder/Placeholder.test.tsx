import * as React from "react";
import { shallow } from "enzyme";
import * as usage from "./Placeholder.usage";
import "jest-styled-components";
import styled from "styled-components";
import { View } from "react-native";
import { Placeholder } from ".";

describe("<Placeholder />", () => {
  for (const example in usage) {
    const x = (usage as any)[example]();
    it(`renders well in ${example} mode`, () => {
      expect(shallow(x)).toMatchSnapshot();
    });
  }

  it(`renders the right orientation`, () => {
    let wrapper = shallow(<Placeholder />);
    expect(wrapper.dive()).toHaveStyleRule("flex-direction", "row");
    wrapper = shallow(<Placeholder orientation="column" />);
    expect(wrapper.dive()).toHaveStyleRule("flex-direction", "column");
  });
});
