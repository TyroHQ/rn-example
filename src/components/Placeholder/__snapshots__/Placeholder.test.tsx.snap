// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Placeholder /> renders well in basic mode 1`] = `
<Styled(Component)
  orientation="row"
>
  <Placeholder
    key="0"
    orientation="row"
  >
    This can be replaced
  </Placeholder>
  <Placeholder
    key="1"
    orientation="row"
  >
    This can be replaced
  </Placeholder>
  <Placeholder
    key="2"
    orientation="row"
  >
    This can be replaced
  </Placeholder>
</Styled(Component)>
`;

exports[`<Placeholder /> renders well in withNumber mode 1`] = `
<Styled(Component)
  orientation="row"
>
  5
</Styled(Component)>
`;

exports[`<Placeholder /> renders well in withString mode 1`] = `
<Styled(Component)
  orientation="row"
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Hello
  </Text>
</Styled(Component)>
`;
